tests/v2/validation/observability/stackstate_test.go

```diff
@@ -0,0 +1,271 @@
+	require.NoError(ss.T(), err)
+
+	var stackstateConfigs observability.StackStateConfigs
+	config.LoadConfig(stackStateConfigFileKey, &stackstateConfigs)
+
+	initialStackstateAgent, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+	require.NoError(ss.T(), err)
+
+	if !initialStackstateAgent.IsAlreadyInstalled {
+		log.Info("Installing stack state agent on the provided cluster")
+
+		systemProject, err := projects.GetProjectByName(client, ss.cluster.ID, systemProject)
+		require.NoError(ss.T(), err)
+		systemProjectID := strings.Split(systemProject.ID, ":")[1]
+
+		err = charts.InstallStackstateAgentChart(ss.client, ss.stackstateAgentInstallOptions, stackstateConfigs.ClusterApiKey, stackstateConfigs.Url, systemProjectID)
+		require.NoError(ss.T(), err)
+		log.Info("Stack state chart installed successfully")
+
+		ss.T().Log("Verifying the deployments of stackstate agent chart to have expected number of available replicas")
+		err = extencharts.WatchAndWaitDeployments(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+		require.NoError(ss.T(), err)
+
+		ss.T().Log("Verifying the daemonsets of stackstate agent chart to have expected number of available replicas nodes")
+		err = extencharts.WatchAndWaitDaemonSets(client, ss.cluster.ID, charts.StackstateNamespace, meta.ListOptions{})
+		require.NoError(ss.T(), err)
+	}
+}
+
+func (ss *StackStateTestSuite) TestUpgradeStackstateAgentChart() {
+	subSession := ss.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := ss.client.WithSession(subSession)
+	require.NoError(ss.T(), err)
+
+	versionsList, err := client.Catalog.GetListChartVersions(charts.StackstateK8sAgent, rancherPartnerCharts)
+	require.NoError(ss.T(), err)
+
+	if len(versionsList) < 2 {
+		ss.T().Skip("Skipping the upgrade case, only one version of stackstate agent chart is available")
+	}
+	require.GreaterOrEqualf(ss.T(), len(versionsList), 2, "There should be at least 2 versions of the stackstate agent chart")
+
+	versionLatest := versionsList[0]
+	ss.T().Log(versionLatest)
+	versionBeforeLatest := versionsList[1]
+	ss.T().Log(versionBeforeLatest)
+	ss.stackstateAgentInstallOptions.Version = versionBeforeLatest
+
+	ss.T().Log("Checking if the stackstate agent chart is installed with one of the previous versions")
+	initialStackstateAgent, err := extencharts.GetChartStatus(client, ss.cluster.ID, charts.StackstateNamespace, charts.StackstateK8sAgent)
+	require.NoError(ss.T(), err)
+
+	if initialStackstateAgent.IsAlreadyInstalled && initialStackstateAgent.ChartDetails.Spec.Chart.Metadata.Version == versionLatest {
+		ss.T().Skip("Skipping the upgrade case, stackstate agent chart is already installed with the latest version")
+	}
+
+	var stackstateConfigs observability.StackStateConfigs
+	config.LoadConfig(stackStateConfigFileKey, &stackstateConfigs)
+
+	if !initialStackstateAgent.IsAlreadyInstalled {
```

Comment:
I'm conflicted about this. I wouldn't think we want this to handle both cases. If it handles both, it opens up the possibility that we will see inconsistencies when executing the test. i.e. if one day I run this with the chart already installed -> upgrade and the upgrade fails, but the test normally passes (because the chart isn't already installed, therefore the test installs it here in a way the test is expecting, then upgrades), this will look 'flaky'. 


I think a better approach would be to have this test install the chart everytime, and fail if its already installed. Then have a separate test that makes sure it _is_ installed, and have a 'dynamicUpgradeChart' test.

Reply:
we can talk more about this in tomorrow's call.